Initial draft of Python 'Security Policy' (#1804)
python/devguide: security/policy.rst

> ===============
> Security Policy
> ===============

Reviewer: Use sentence case for headers:

https://devguide.python.org/documentation/style-guide/#capitalization

```suggestion
===============
Security policy
===============
```

> Python Security Response Team (PSRT) members balance this work against

Reviewer 2: Can we please add a link to the PSRT page?

many other responsibilities. Please be thoughtful about the time and attention
your report requires.  Repeated failure to respect the security policy will
result in future reports being rejected or being banned from the ``python``
GitHub organization, regardless of technical merit.

What types of bugs are vulnerabilities?
---------------------------------------

Not all bugs are vulnerabilities. To avoid causing
duplicate work for PSRT members all potential reports
must be evaluated against the relevant threat models
prior to being submitted to the PSRT.
Where possible, cite the relevant threat model to show that
the threat model has been considered while determining whether
to report a bug as a vulnerability.

Vulnerabilities must be exploitable from code, configurations,
pre-conditions, and deployments that might feasibly exist in
the real world. For example, a vulnerability only affecting code
that does not make sense in a production program
will not be accepted as a vulnerability.

Documented functionality will not be considered a vulnerability.
For example, :mod:`pickle`, :mod:`marshal`, :mod:`shelve`, :mod:`eval`,

> and :mod:`exec` are documented to execute arbitrary Python code that is supplied

Reviewer: Fix links:

```suggestion
For example, :mod:`pickle`, :mod:`marshal`, :mod:`shelve`, :func:`eval`,
and :func:`exec` are documented to execute arbitrary Python code that is supplied
```

as data. The :mod:`ctypes` module is documented to enable modifying arbitrary
locations in memory.

Vulnerabilities must not depend on malicious control of Python's launch
conditions, including command line arguments, environment variables, or
modifications to files on the target system. We assume that, at the time Python
is executed, the environment is as intended by the legitimate user, and any
malicious variation from this cannot be mitigated by Python itself.

Vulnerabilities that affect availability (such as DoS, ReDoS) must be
triggerable with data inputs that are reasonably sized for the use-case.
Availability vulnerabilities must also demonstrate an "upward" change in posture
for the attacker, rather than a "lateral" change in posture.
This is to avoid handling performance improvements as security vulnerabilities.

Vulnerabilities in dependencies of Python (such as zlib, Tcl/Tk, or OpenSSL)
are not vulnerabilities in Python unless Python's use of the dependency
interferes with secure use of the dependency.
For example, a vulnerability in the bundled copy of zlib in Python is a
vulnerability in zlib, not Python.

What versions of Python are accepting reports?

> ----------------------------------------------

Reviewer: ```suggestion
What versions of Python accept reports?
---------------------------------------
```

> Python accepts security vulnerabilities and will

Reviewer: Let's accept _reports_ rather than _vulnerabilities_ 🙃
```suggestion
Python accepts vulnerability reports and will
```

assign CVE IDs for `supported Python versions <branchstatus>`_ that have a status of
`"bugfix" or "security" <version-status-key>`_. Versions that are not yet

> stable (status of `"feature" or "prerelease" <version-status-key>`_) are not

Reviewer: Prefer `:ref:` for internal references https://devguide.python.org/documentation/markup/#quick-reference

```suggestion
assign CVE IDs for :ref:`supported Python versions <branchstatus>` that have a status of
:ref:`"bugfix" or "security" <version-status-key>`. Versions that are not yet
stable (status of :ref:`"feature" or "prerelease" <version-status-key>`) are not
```

  or requesting compliance or due-diligence work)
  will be discarded without a reply.

How to submit a vulnerability report?
-------------------------------------

Submit all potential security vulnerability reports for CPython
to GitHub Security Advisories
by `opening a new ticket <GHSA>`__.
Do not open a public GitHub issue to report a security vulnerability.
For all other projects (such as pip, python.org and tools) or if you're
not sure where to send your report, send an email to
`security@python.org <mailto:security@python.org>`__.

Here's what to expect for how a vulnerability report will be handled:

* Reporter reports the vulnerability privately to the PSRT.
* If the PSRT determines the report isn't a vulnerability, the issue
  can be opened in the public issue tracker.
* If the PSRT determines the report is a vulnerability, the PSRT will
  accept your report and a CVE ID will be assigned by the PSF CNA.
* Once a public pull request containing a fix is merged to CPython,
  the advisory and CVE record will be published with attribution.

For more information about how the PSRT handles vulnerabilities,

> `consult the Python Developer Guide <psrt-vulnerability-process>`__.

Reviewer 2: We are in the devguide here.

Reviewer: ```suggestion
see :ref:`psrt-vulnerability-process`.
```

PSF Code of Conduct

> -------------------

Reviewer: ```suggestion
Code of conduct
---------------
```

===============
Security policy
===============

:ref:`Python Security Response Team <psrt>` (PSRT) members balance this work against
many other responsibilities. Please be thoughtful about the time and attention

> your report requires.  Repeated failure to respect the security policy will

Reviewer 3: ```suggestion
your report requires. Repeated failure to respect the security policy will
```

> result in future reports being rejected or being banned from the ``python``

Reviewer 3: ```suggestion
result in future reports being rejected or banned from the ``python``
```

GitHub organization, regardless of technical merit.

What types of bugs are vulnerabilities?
---------------------------------------

Not all bugs are vulnerabilities. To avoid causing

> duplicate work for PSRT members all potential reports

Reviewer 3: ```suggestion
duplicate work for PSRT members, all potential reports
```

must be evaluated against the relevant threat models
prior to being submitted to the PSRT.
Where possible, cite the relevant threat model to show that

> the threat model has been considered while determining whether

Reviewer 3: ```suggestion
the latter has been considered while determining whether
```

to report a bug as a vulnerability.

Vulnerabilities must be exploitable from code, configurations,

> pre-conditions, and deployments that might feasibly exist in

Reviewer 3: ```suggestion
pre-conditions, or deployments that might feasibly exist in
```

> pre-conditions, and deployments that might feasibly exist in

Reviewer 3: "might feasibly exist" seems overly redundant and I would have just said "may exist". We already say "in the real world" later so this should be enough.

the real world. For example, a vulnerability only affecting code
that does not make sense in a production program
will not be accepted as a vulnerability.

> Documented functionality will not be considered a vulnerability.

Reviewer 3: ```suggestion
Documented functionality is not considered a vulnerability.
```

There are places where I would use the present instead of a future form.

For example, :mod:`pickle`, :mod:`marshal`, :mod:`shelve`, :func:`eval`,
and :func:`exec` are documented to execute arbitrary Python code that is
supplied as data. The :mod:`ctypes` module is documented to enable modifying
arbitrary locations in memory.

Vulnerabilities must not depend on malicious control of Python's launch

> conditions, including command line arguments, environment variables, or

Reviewer 3: ```suggestion
conditions, including (but not limited to) command line arguments, environment variables, or
```

So that LLMs parsing the text know that these are not the only conditions. You can ignore this amendment though.

modifications to files on the target system. We assume that, at the time Python
is executed, the environment is as intended by the legitimate user, and any
malicious variation from this cannot be mitigated by Python itself.

> Vulnerabilities that affect availability (such as DoS, ReDoS) must be

Reviewer 3: ```suggestion
Vulnerabilities that affect availability (such as DoS or ReDoS) must be
```

triggerable with data inputs that are reasonably sized for the use-case.
Availability vulnerabilities must also demonstrate an "upward" change in posture

> for the attacker, rather than a "lateral" change in posture.

Reviewer 3: ```suggestion
for the attacker, rather than a "lateral" one.
```

To avoid repetitions

These features are not eligible for security vulnerabilities.
Instead open a public GitHub issue.

If a vulnerability is platform-dependent, check if the platform is
supported per :pep:`11`.
Vulnerabilities that exclusively affect unsupported platforms
may not be accepted.

What to include and how to structure a vulnerability report?
------------------------------------------------------------

For your vulnerability report to be handled efficiently by
the PSRT, the report must include certain information and
be formatted correctly:

* For the initial report and follow-up communications, avoid
  overly long, verbose, or excessive structure (such as headers or tables).
  Ideally reports should be a few sentences describing the vulnerability and
  a proof-of-concept script that reproduces the issue and provides a clear
  indication whether the vulnerability is still present (such as exiting with
  ``1`` if vulnerable and ``0`` if not vulnerable).
* When reporting large numbers or "batches" of vulnerabilities or
  searching for potential vulnerabilities using an LLM, you as a reporter must
  verify the validity of all reports prior to submission to the PSRT.

>   PSRT members WILL NOT spend time confirming the validity of reports, only

Reviewer 3: If you want to stress things using capitalization like in RFCs, I would suggest doing the same in other places where we say MUST or MAY.

Author: Done in https://github.com/python/devguide/pull/1804/commits/2d403e1d7a5f525ed2ecfea946b7b7af15bd4250

  whether a valid bug report is a vulnerability or not.
* Do not include severity or CVSS information in your initial report,
  this information will be determined by the PSRT.
* Ideally, include a minimal patch with the mitigation for the report.
* Always include the versions of Python that were tested,
  and indicate which were found to be vulnerable.

> * Submit reports as plain-text only, including attachments.

Reviewer 3: Can we tell them to submit PoCs in `<details>` tags as well? I hate downloading stuff even if it is a .py script. In particular I prefer not having any attachment at all.

Author: Done in https://github.com/python/devguide/pull/1804/commits/2d403e1d7a5f525ed2ecfea946b7b7af15bd4250

  No PDFs, binaries, notebooks, or other files that cannot be safely reviewed.
  If your proof-of-concept depends on a specially constructed binary file,
  please include a script to construct it rather than the file itself.
* Reports that do not contain a potential security vulnerability (such as spam
  or requesting compliance or due-diligence work)
  will be discarded without a reply.

How to submit a vulnerability report?
-------------------------------------

Submit all potential security vulnerability reports for CPython
by `opening a new ticket <GHSA>`__.
Do not open a public GitHub issue to report a security vulnerability.
For all other projects (such as pip, python.org and tools) or if you're
not sure where to send your report, send an email to
`security@python.org <mailto:security@python.org>`__.

Here's what to expect for how a vulnerability report will be handled:

* Reporter reports the vulnerability privately to the PSRT.
* If the PSRT determines the report isn't a vulnerability, the issue
  can be opened in the public issue tracker.
* If the PSRT determines the report is a vulnerability, the PSRT will

>   accept your report and a CVE ID will be assigned by the PSF CNA.

Reviewer 3: The first bullet point mentions "Reporter" as an impersonal entity but here it becomes a "you". It should be consistent (e.g. "the report")

Author: Done in https://github.com/python/devguide/pull/1804/commits/2d403e1d7a5f525ed2ecfea946b7b7af15bd4250

===============
Security policy
===============

> :ref:`Python Security Response Team <psrt>` (PSRT) members balance this work against

Reviewer 4: Perhaps make this more prominent by adding a directive.

many other responsibilities. Please be thoughtful about the time and attention
your report requires. Repeated failure to respect the security policy will
result in future reports being rejected or banned from the ``python``

> GitHub organization, regardless of technical merit.

Reviewer 5: ```suggestion
your report requires. Repeated failure to respect the security policy will
result in future reports being rejected, or the reporter being banned from the
``python`` GitHub organization, regardless of technical merit.
```

I think we dropped a noun here?

What types of bugs are vulnerabilities?
---------------------------------------

> Not all bugs are vulnerabilities. To avoid causing

Reviewer 4: Offset the first sentence and bold or highlight.

duplicate work for PSRT members, all potential reports
must be evaluated against the relevant threat models
prior to being submitted to the PSRT.
Where possible, cite the relevant threat model to show that
the latter has been considered while determining whether

> to report a bug as a vulnerability.

Reviewer 4: ```suggestion
**Not all bugs are vulnerabilities.**

Before submitting a report to the PSRT, **all potential** reports
must be evaluated against the relevant threat models.
Cite the relevant threat model if reporting a bug as a vulnerability.
```

Vulnerabilities must be exploitable from code, configurations,
pre-conditions, or deployments that may exist in the real world.
For example, a vulnerability only affecting code
that does not make sense in a production program

> will not be accepted as a vulnerability.

Reviewer 4: ```suggestion
Vulnerabilities must be exploitable from code, configurations,
pre-conditions, or deployments that exist in the real world.
A vulnerability that only affects code
unlikely to be used in a production program
will not be accepted.
```

Documented functionality is not considered a vulnerability.
For example, :mod:`pickle`, :mod:`marshal`, :mod:`shelve`, :func:`eval`,
and :func:`exec` are documented to execute arbitrary Python code that is
supplied as data. The :mod:`ctypes` module is documented to enable modifying
arbitrary locations in memory.

Vulnerabilities must not depend on malicious control of Python's launch
conditions, including (but not limited to) command line arguments, environment variables, or
modifications to files on the target system. We assume that, at the time Python
is executed, the environment is as intended by the legitimate user, and any
malicious variation from this cannot be mitigated by Python itself.

Vulnerabilities that affect availability (such as DoS, ReDoS, crashes,
dead-locks, and resource exhaustion) must be

> triggerable with data inputs that are reasonably sized for the use-case.

Reviewer 5: ```suggestion
triggerable with data inputs that are reasonably sized for the use case.
```

Availability vulnerabilities must also demonstrate an "upward" change in posture

> for the attacker, rather than a "lateral" one.

Reviewer 6: This is probably clear to security experts, but I'm not familiar with the terms. Could you add an explanation or link?

Author: cc @woodruffw if you've got a better way to phrase this or a resource for the terminology? I am taking this to mostly mean that an availability vulnerability can't be: "attacker can send a lot of data and that makes the operation take longer" but can be: "attacker can send an input that crashes the program".

Reviewer 7: That terminology is what I've seen before! Another way to phrase it is that the attacker starts with certain capabilities, and in order for something to be considered vulnerable it needs to reward the attacker with _new_ capabilities.

Vulnerabilities in dependencies of Python (such as zlib, Tcl/Tk, or OpenSSL)
are not vulnerabilities in Python unless Python's use of the dependency
interferes with secure use of the dependency.
For example, a vulnerability in the bundled copy of zlib in Python is a
vulnerability in zlib, not Python.

What versions of Python accept reports?
---------------------------------------

Python accepts vulnerability reports and will
assign CVE IDs for :ref:`supported Python versions <branchstatus>` that have a
status of :ref:`"bugfix" or "security" <version-status-key>`. Versions that are
not yet stable (status of :ref:`"feature" or "prerelease" <version-status-key>`)
are not eligible for CVE IDs. If the vulnerability only exists in prerelease
versions (alphas, betas, release candidates), then the issue should be reported
as a regular bug.
Prior to submitting a report, check whether the issue has already been
resolved on the ``main`` branch and only requires backporting.

Sometimes features may be marked as
"experimental" in Python, even in a stable Python version.
These features are not eligible for security vulnerabilities.

> Instead open a public GitHub issue.

Reviewer 5: ```suggestion
Instead, open a public GitHub issue.
```

If a vulnerability is platform-dependent, check if the platform is
supported per :pep:`11`.
Vulnerabilities that exclusively affect unsupported platforms

> may not be accepted.

Reviewer 6: We want to wash our hands, but IMO a `pyoton/cpython` GHSA is often the best place to coordinate.
In practice these are corporate projects (e.g. Solaris) or hobbyists (who should *know* but can't provide guarantees).

Would something like this make sense?

```suggestion
Vulnerabilities that exclusively affect unsupported platforms
are not treated as vulnerabilities in Python.

As per the :pep:`Unsupported Platforms section of PEP 11 <11#unsupported-platforms>`,
porting Python to such a platform is treated as a third-party project.
If you choose to report such a vulnerability to Python, please follow the
requirements of this guide, but note that the report may be shared with
parties who expressed interested in the relevant platforms, and will
generally be handled according to the relevant maintainers' security
policies, or closed if the maintainers are unknown or unresponsive.
```

Feel free to assign these to me if any comes up.
For my info: are we allowed to assign a CVE number for such a "third-party" port?

Author: If the fix lands in CPython itself then there's no ambiguity, we get to decide. If the fix doesn't land in CPython I think the story is not as simple. I think for now we should assume that we wouldn't assign a CVE ID in this case, but we can just as easily go to Red Hat or MITRE for a CVE ID in these cases.

What to include and how to structure a vulnerability report?
------------------------------------------------------------

For your vulnerability report to be handled efficiently by
the PSRT, the report must include certain information and
be formatted correctly:

* For the initial report and follow-up communications, avoid
  overly long, verbose, or excessive structure (such as headers or tables).
  Ideally reports should be a few sentences describing the vulnerability and
  a proof-of-concept script that reproduces the issue and provides a clear

>   indication whether the vulnerability is still present (such as exiting with

Reviewer 5: ```suggestion
  indication of whether the vulnerability is still present (such as exiting with
```

>   ``1`` if vulnerable and ``0`` if not vulnerable).

Reviewer 4: ```suggestion
  Reports should be a few sentences describing the vulnerability. Ideally, include
  a proof-of-concept script that reproduces the issue and provides a clear
  indication of whether the vulnerability is still present (such as exiting with
  ``1`` if vulnerable and ``0`` if not vulnerable).
```

  please include a script to construct it rather than the file itself.
* Proof-of-concept scripts longer than a few lines should be wrapped with a
  `collapsed section`_ using ``<details></details>`` for better readability.
* Reports that do not contain a potential security vulnerability (such as spam
  or requesting compliance or due-diligence work)
  will be discarded without a reply.

.. _collapsed section: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections

How to submit a vulnerability report?
-------------------------------------

Submit all potential security vulnerability reports for CPython
by `opening a new ticket <GHSA>`__.
Do not open a public GitHub issue to report a security vulnerability.
For all other projects (such as pip, python.org and tools) or if you're
not sure where to send your report, send an email to
`security@python.org <mailto:security@python.org>`__.

Here's what to expect for how a vulnerability report will be handled:

* Reporter reports the vulnerability privately to the PSRT.
* If the PSRT determines the report isn't a vulnerability, the issue

>   can be opened in the public issue tracker.

Reviewer 5: ```suggestion
* If the PSRT determines the report isn't a vulnerability, the reporter
   may open a public issue.
```

  No PDFs, binaries, notebooks, or other files that cannot be safely reviewed.
  If your proof-of-concept depends on a specially constructed binary file,
  please include a script to construct it rather than the file itself.
* Proof-of-concept scripts longer than a few lines should be wrapped

>   with ``<detail></detail>`` for better readability.

Reviewer 3: I think it's `<details>` (sorry if I mispelled it)

Reviewer 2: Also, you can link to https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections.
